cirq/contrib/two_qubit_gates/gate_compilation.py

```diff
@@ -0,0 +1,328 @@
+    # Index j corresponds to KAK_vecs[j], and is of the form
+    # ( (u0[0],u1[0]), (u0[1],u1[1]), ...) where u0[k] is the kth single qubit
+    # unitary acting on qubit 0 (similarly for u1)
+    single_qubit_gates: Sequence[Sequence[_SingleQubitGatePair]]
+    max_expected_infidelity: float
+
+    def compile_two_qubit_gate(
+            self, unitary: np.ndarray
+    ) -> Tuple[Sequence[_SingleQubitGatePair], np.ndarray, bool]:
+        r"""Compute single qubit gates required to compile a desired unitary.
+
+        Given a desired unitary U, this computes the sequence of 1-local gates
+        k_j such that the product
+
+        k_{n-1} A k_{n-2} A ... k_1 A k_0
+
+        is close to U. Here A is the base_gate of the tabulation.
+
+        Args:
+            unitary: Unitary (U above) to compile.
+
+        Returns:
+            local_unitaries: Sequence of 2-tuples
+                (k_{00},k_{01}), (k_{10},k_{11}) ...
+                where k_j = k_{j0}\otimes k_{j1} in the product above.
+            actual_unitary: The actual outcome of the sequence of products
+                above.
+            success: Whether the actual_unitary is expected to be within the
+                the desired maximum entanglement infidelity to the target
+                unitary.
+        """
+        unitary = np.asarray(unitary)
+        kak_vec = kak_vector(unitary, check_preconditions=False)
+        infidelities = KAK_vector_infidelity(kak_vec,
+                                             self.KAK_vecs,
+                                             ignore_equivalent_vectors=True)
+        nearest_ind = infidelities.argmin()
+
+        success = infidelities[nearest_ind] < self.max_expected_infidelity
+
+        # shape (n,2,2,2)
+        inner_gates = np.array(self.single_qubit_gates[nearest_ind])
+
+        if inner_gates.size == 0:  # Only need base gate
+            outer_locals, actual = _outer_locals_for_unitary(
+                unitary, self.base_gate)
+            return outer_locals, actual, success
+
+        # reshape to operators on 2 qubits, (n,4,4)
+        inner_gates = vector_kron(inner_gates[..., 0, :, :],
+                                  inner_gates[..., 1, :, :])
+
+        assert inner_gates.ndim == 3
+        inner_product = reduce(lambda a, b: self.base_gate @ b @ a, inner_gates,
+                               self.base_gate)
+        outer_locals, actual = _outer_locals_for_unitary(unitary, inner_product)
+
+        out = [outer_locals[0]]
+        out.extend(self.single_qubit_gates[nearest_ind])
+        out.append(outer_locals[1])
+
+        return out, actual, success
+
+
+def _outer_locals_for_unitary(
+        target: np.ndarray, base: np.ndarray
+) -> Tuple[Tuple[_SingleQubitGatePair, _SingleQubitGatePair], np.ndarray]:
+    """Local unitaries mapping between locally equivalent 2-local unitaries.
+
+    Finds the left and right 1-local unitaries k_L, k_R such that
+
+    U_target = k_L @ U_base @ k_R
```

Comment:
Do we need the `k_`? Why not just `L` and `R`?

Reply:
Done